### PR TITLE
config(cosmiconfig): allow rules passed in config files

### PR DESCRIPTION
**Pull request by goldcaddy77**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -37,6 +37,7 @@
     "chalk": "^2.0.1",
     "columnify": "^1.5.4",
     "commander": "^2.11.0",
+    "cosmiconfig": "davidtheclark/cosmiconfig#3.0",
```
Comment (goldcaddy77) [inline package.json]: Pinned at 3.0 with the `sync` changes.

File: src/cli.js [hunk 2]
```diff
@@ -11,7 +11,7 @@ process.on('uncaughtException', err => {
     'It would be super helpful if you could report this here: https://github.com/cjoudrey/graphql-schema-linter/issues/new'
   );
   console.error('');
-  console.error(err);
+  console.error(err.stack);
```
Comment (goldcaddy77) [inline src/cli.js]: The standard error message doesn't give much info.  Thoughts on showing the full stack trace?

File: src/configuration.js [hunk 3]
```diff
@@ -32,30 +35,42 @@ export class Configuration {
     }
   }
 
-  getRules() {
+  getRules(searchDirectory) {
```
Comment (cjoudrey) [inline src/configuration.js]: Rather than accept a `searchDirectory` here, what are your thoughts on having the `constructor` accept a `configDirectory` argument?

My reasoning is that eventually other things will be part of this config file and we'd have to accept a `searchDirectory` for all of them. For example, `.graphqlconfig` supports specifying where your `schema` is located. We could leverage this information in `getSchema()` so that people can just run `graphql-schema-linter` with no options at all.

Comment (goldcaddy77) [inline src/configuration.js]: Good thinking 👍

File: src/configuration.js [hunk 4]
```diff
@@ -36,26 +39,38 @@ export class Configuration {
...
     } else if (this.options.except.length > 0) {
       rules = defaultRules.filter(rule => {
         return (
           this.options.except.map(toUpperCamelCase).indexOf(rule.name) == -1
         );
       });
     } else {
-      rules = defaultRules;
+      const directory = this.options.searchDirectory || process.cwd();
+      const cosmic = cosmiconfig('graphql-schema-linter', {
```
Comment (cjoudrey) [inline src/configuration.js]: Along the same line of thinking, we might want to move this to a `func loadOptionsFromConfig()` that gets called in the `constructor` otherwise we will have this same boilerplate in `getSchema`.

We could also do something like this in the `constructor`:

```js
constructor(options, stdinFd, configSearchDirectory) {
  const configOptions = this.loadOptionsFromConfig(configSearchDirectory);
  this.options = Object.assign(configOptions, options);
}
```

This way CLI options take precedence over config. It also enables us to easily add support for [`.graphqlconfig`](https://github.com/graphcool/graphql-config) in the future.

I think the beauty of doing this in the `constructor` is that none of the getters need to be changed.

Comment (goldcaddy77) [inline src/configuration.js]: Cool, will make the change.

File: src/configuration.js [hunk 5]
```diff
@@ -1,8 +1,11 @@
...
+import path from 'path';
+
 import defaultRules from './rules/index.js';
 import JSONFormatter from './formatters/json_formatter.js';
 import TextFormatter from './formatters/text_formatter.js';
-import getGraphQLProjectConfig from 'graphql-config';
-import { readSync, readFileSync } from 'fs';
 
 export class Configuration {
   constructor(options, stdinFd) {
```
Comment (goldcaddy77) [inline src/configuration.js]: Think I should define the config options here in a comment?

```js
  /*
    options:
      - format: (required) `text` | `json`
      - args: command line args (arg[0] = schema filename)
      - except: [comma-delimited list] blacklist rules
      - only: [comma-delimited list] whitelist rules
      - searchDirectory: [path] path for where to start looking for config
      - stdin: [boolean] (pass schema via stdin)
  */
```

File: src/cli.js (hunk 2 above)
Comment (cjoudrey) [inline src/cli.js]: Oops! I was under the impression the full thing was getting logged. If `err.stack` contains more info then :+1: :+1:.

File: src/configuration.js (hunk 5 above)
Comment (cjoudrey) [inline src/configuration.js]: I guess that's the disadvantage of not using something like TypeScript. Maybe in the future we could port it over.

One thing to keep in mind is that `except` and `only` are both arrays at this point since the transformation is done here: https://github.com/goldcaddy77/graphql-schema-linter/blob/68a5c6916328a82c19553e2677d40a771240dc79/src/runner.js#L8

Maybe we should also :fire: `args` and have `schemaPath` that way the `runner.js` does the `args` handling.

I wonder if `searchDirectory` belongs in this list though. All the other options are _actual_ options. I kinda feel like it should be `constructor(options, stdinFd, searchDirectory)`.

That way `options` only contains the things you can actually specify via CLI options or configuration file.

**Comment by cjoudrey** [comment]
This is super cool btw thanks for the contribution. 😁 

Sorry for all the comments, by adding "load options from a config file" a lot of things started to be clearer and some things I had done previously started to make less sense. 😆

**Comment by goldcaddy77** [comment]
No worries at all - these comments all make sense. I like the idea of settings pulling from both command line ages and config settings, with the command line given preference. Will get these changes in.

**Comment by goldcaddy77** [comment]
@cjoudrey made a bunch of updates in the last commit and commented inline: https://github.com/cjoudrey/graphql-schema-linter/pull/26/commits/2d9e35b23a67c5868aaa7b5ea3181fbe7aa516cd

File: src/configuration.js [hunk 6]
```diff
@@ -8,8 +8,22 @@ import JSONFormatter from './formatters/json_formatter.js';
...
+      - schemaFileName: [string] file to read schema from
+      - stdin: [boolean] pass schema via stdin?
+  */
+  constructor(options = {}, stdinFd = null) {
+    const defaultOptions = { format: 'text' };
+    const configOptions = loadOptionsFromConfig(
+      options.configDirectory || process.cwd()
+    );
+
+    this.options = Object.assign({}, defaultOptions, configOptions, options);
```
Comment (goldcaddy77) [inline src/configuration.js]: Set up the waterfall - command line > config file > defaults.

File: src/configuration.js [hunk 7]
```diff
@@ -41,30 +55,35 @@ export class Configuration {
 
     var rules = defaultRules;
 
-    if (this.options.only.length > 0) {
+    if (this.options.only && this.options.only.length > 0) {
       rules = filterRules(this.options.only);
-    } else if (this.options.except.length > 0) {
+    } else if (this.options.except && this.options.except.length > 0) {
```
Comment (goldcaddy77) [inline src/configuration.js]: `only` and `except` no longer required.

File: src/configuration.js [hunk 8]
```diff
@@ -41,30 +55,35 @@ export class Configuration {
...
         return (
           this.options.except.map(toUpperCamelCase).indexOf(rule.name) == -1
         );
       });
-    } else {
-      const directory = this.options.searchDirectory || process.cwd();
-      const cosmic = cosmiconfig('graphql-schema-linter', {
-        cache: false,
-        sync: true,
-      }).load(directory);
```
Comment (goldcaddy77) [inline src/configuration.js]: You were right... much cleaner now 👍

File: src/configuration.js [hunk 9]
```diff
@@ -41,30 +55,35 @@ export class Configuration {
...
 }
 
+function loadOptionsFromConfig(directory) {
+  const cosmic = cosmiconfig('graphql-schema-linter', {
+    cache: false,
+    sync: true,
+  }).load(directory);
+  let options = {};
+  if (cosmic) {
+    // Map config.rules -> `only` param
```
Comment (goldcaddy77) [inline src/configuration.js]: Most linters use a `rules` key, so I've adopted that and mapped it to `only`.  Perhaps in a future version we could bump the major version number, drop `except` and change the command line flag to `rules`?


- http://eslint.org/docs/user-guide/configuring#extending-configuration-files
- https://palantir.github.io/tslint/usage/configuration/

File: src/runner.js [hunk 10]
```diff
@@ -12,6 +12,10 @@ export function run(stdout, stdin, argv) {
       'only the rules specified will be used to validate the schema. Example: FieldsHaveDescriptions,TypesHaveDescriptions'
     )
     .option(
+      '-c, --config-directory <path>',
+      'path to begin searching for config files.'
+    )
```
Comment (goldcaddy77) [inline src/runner.js]: I think this makes sense to expose to command line.

File: src/runner.js [hunk 11]
```diff
@@ -47,3 +45,29 @@ export function run(stdout, stdin, argv) {
 
   return errors.length > 0 ? 1 : 0;
 }
+
+function getOptionsFromCommander(commander) {
```
Comment (goldcaddy77) [inline src/runner.js]: This way, we only pass in the command line flags that are actually filled in.

File: src/configuration.js (hunk 9 above)
Comment (cjoudrey) [inline src/configuration.js]: Yeah, the more I think about it the more I agree with you. I'd be down to remove `--except` and rename `--only` to `--rules`. If it's easier we can do it as a separate PR, otherwise we can do it in this one. :+1:

File: src/runner.js (hunk 10 above)
Comment (cjoudrey) [inline src/runner.js]: Would love to hear your thoughts on this one.

I expected most people would run `graphql-schema-linter` from within their project's folder. Even if they aren't at the root (i.e. `/home/cjoudrey/my-project/tests/something`), `cosmiconfig` will move up directories until it finds a config or until it reaches home directory `~/`.

File: src/configuration.js [hunk 12]
```diff
@@ -1,12 +1,29 @@
...
+      - except: [string array] blacklist rules
+      - format: (required) `text` | `json`
+      - only: [string array] whitelist rules
+      - schemaFileName: [string] file to read schema from
+      - stdin: [boolean] pass schema via stdin?
+  */
+  constructor(options = {}, stdinFd = null) {
+    const defaultOptions = { format: 'text' };
+    const configOptions = loadOptionsFromConfig(
+      options.configDirectory || process.cwd()
```
Comment (cjoudrey) [inline src/configuration.js]: Rather than use `process.cwd()` do you think we should make this class completely agonistic to the env?

i.e.

In `src/cli.js` we'd do:

```diff
- const exitCode = run(process.stdout, process.stdin, process.argv);
+ const exitCode = run(process.stdout, process.stdin, process.argv, process.cwd());
```

and `src/configuration.js`:

```diff
- constructor(options = {}, stdinFd = null) {
+ constructor(options = {}, stdinFd = null, currentWorkingDir) {
```

My goal was to keep all the env-related things in `cli.js` that way all the other classes can be easily tested and do not rely on `process`. Thoughts?

File: src/configuration.js (hunk 9 above)
Comment (goldcaddy77) [inline src/configuration.js]: Awesome, yeah I think that will be cleaner.  I think we should do this in another PR for 2 reasons:

1. this is becoming a monster already 😆 
2. this would be a big breaking change and we should likely bump the major version number.

File: src/runner.js (hunk 10 above)
Comment (goldcaddy77) [inline src/runner.js]: I completely agree with you that this will be 99% of the use cases.  I'm actually working on another project where we need to be able to supply the location of where the config files are, though, so I figured others might need this, too.  Are you cool with keeping?

File: src/configuration.js (hunk 12 above)
Comment (goldcaddy77) [inline src/configuration.js]: So it turns out you can either pass a config directory (full path) or a search path (relative to `process.cwd()`), so I don't even need to pass process.cwd() all the way down.  See: https://github.com/davidtheclark/cosmiconfig/blob/master/lib/createExplorer.js#L52

File: src/configuration.js [hunk 13]
```diff
@@ -36,26 +51,47 @@ export class Configuration {
...
     }
 
     return rules;
   }
 }
 
+function loadOptionsFromConfig(configDirectory) {
+  // If config path is not specified, look in root directory of project
+  // the first option to cosmiconfig.load can be absolute or relative
+  const searchPath = configDirectory || './'
```
Comment (goldcaddy77) [inline src/configuration.js]: @cjoudrey - much cleaner now.

File: README.md [hunk 14]
```diff
@@ -59,6 +59,36 @@ Options:
     output usage information
 ```
 
+## Config
+
+In addition to being able to spefify rules on the command line, config can also be placed in the following locations:
```
Comment (cjoudrey) [inline README.md]: `spefify` => `specify`

File: src/runner.js (hunk 10 above)
Comment (cjoudrey) [inline src/runner.js]: I usually side on YAGNI 😆 in this case, but if you need it, we can add it.

**Comment by cjoudrey** [comment]
Thanks for the contribution! 😁
